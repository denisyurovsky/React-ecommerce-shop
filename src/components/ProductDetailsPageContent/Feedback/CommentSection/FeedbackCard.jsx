import { KeyboardArrowUp } from '@mui/icons-material';
import { Card, CardContent, Rating, Typography, Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { MAX_SHOWN_SYMBOLS } from '../../../../helpers/constants/feedbackConstants';
import { formatDate } from '../../../../helpers/utils/formatData';

import styles from './feedbackCard.module.scss';

const FeedbackCard = ({ review }) => {
  const { comment, createdAt, rating, displayedName } = review;
  const [isOpen, setIsOpen] = useState(false);
  const onClick = () => setIsOpen(!isOpen);

  const date = formatDate(createdAt);
  const isLong = comment.length > MAX_SHOWN_SYMBOLS;
  const isLimited = isLong && !isOpen;
  const content = isLimited
    ? `${comment.slice(0, MAX_SHOWN_SYMBOLS)}...`
    : comment;

  return (
    <Card sx={{ mb: 3 }}>
      <CardContent className={styles.comment}>
        <Typography variant="h5" sx={{ mb: 2, ml: 1, mt: 1 }}>
          {displayedName} ({date})
        </Typography>
        <Rating sx={{ mb: 4, ml: 1 }} value={rating} readOnly size="small" />
        {isLong ? (
          <Typography variant="body2">
            {content}
            <Button
              data-testid="toggleVisibility"
              size="small"
              onClick={onClick}
              className={`${styles.btn} ${isOpen ? styles.bottom : ''}`}
            >
              {isOpen ? <KeyboardArrowUp /> : 'Read more'}
            </Button>
          </Typography>
        ) : (
          <Typography variant="body2">{content}</Typography>
        )}
      </CardContent>
    </Card>
  );
};

FeedbackCard.propTypes = {
  review: PropTypes.shape({
    createdAt: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    displayedName: PropTypes.string.isRequired,
    comment: PropTypes.string.isRequired,
  }),
};

export default FeedbackCard;
