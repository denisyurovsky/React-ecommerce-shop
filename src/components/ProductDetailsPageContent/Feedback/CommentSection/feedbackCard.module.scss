.comment {
  position: relative;
}

.btn {
  text-transform: "none";
  padding-top: 0;
  padding-bottom: 0;
  margin-left: 8px;
}

.bottom {
  position: absolute !important;
  bottom: 6px;
  right: 10px;
}
