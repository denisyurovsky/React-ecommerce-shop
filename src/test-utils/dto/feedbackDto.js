const feedbackDto = [
  {
    id: 0,
    userId: 1,
    productId: 2,
    rating: 5,
    comment: 'Cool stuff',
    createdAt: '2021-12-24T07:49:50.440Z',
    displayedName: 'Fedor',
    productName: 'Headphones',
  },
  {
    id: 1,
    userId: 2,
    productId: 2,
    rating: 3,
    comment: 'Awful thing',
    createdAt: '2021-12-28T13:57:12.255Z',
    displayedName: 'Petr',
  },
  {
    id: 2,
    userId: 3,
    productId: 1,
    rating: 4,
    comment:
      'A very long comment to test read more functionality In mollit laborum cillum qui sit in. Nostrud adipisicing deserunt reprehenderit anim elit aute dolore minim aliqua. Elit magna sunt nisi officia non quis proident excepteur culpa duis. Consequat laborum et anim consequat nulla enim cupidatat laboris duis excepteur eu mollit laboris incididunt. Exercitation amet eiusmod voluptate ad incididunt minim exercitation exercitation laboris sint.',
    createdAt: '2021-12-25T13:56:01.736Z',
    displayedName: 'Anonymous',
    productName: 'Headphones',
  },
];

export default feedbackDto;
