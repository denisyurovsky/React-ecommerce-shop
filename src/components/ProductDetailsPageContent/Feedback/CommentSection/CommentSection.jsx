import MoodIcon from '@mui/icons-material/Mood';
import { Typography, Container, Box } from '@mui/material';
import PropTypes from 'prop-types';
import React from 'react';

import FeedbackCard from './FeedbackCard';

const CommentSection = ({ comments }) => {
  const isEmpty = comments.length === 0;

  return (
    <Box data-testid="comments">
      {isEmpty ? (
        <Container sx={{ display: 'flex', justifyContent: 'center' }}>
          <Typography variant="subtitle1">
            There are no comments yet. Be the first to review
          </Typography>
          <MoodIcon sx={{ ml: 1 }} color="primary" />
        </Container>
      ) : (
        comments.map((review) => (
          <FeedbackCard key={review.id} review={review} />
        ))
      )}
    </Box>
  );
};

CommentSection.propTypes = {
  comments: PropTypes.array.isRequired,
};

export default CommentSection;
