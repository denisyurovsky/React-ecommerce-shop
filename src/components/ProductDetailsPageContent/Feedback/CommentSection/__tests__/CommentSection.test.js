import { render } from '@testing-library/react';
import React from 'react';

import feedbackDto from '../../../../../test-utils/dto/feedbackDto';
import CommentSection from '../CommentSection';

describe('empty feedbacks section', () => {
  it('should show correct message', () => {
    const { getByText } = render(<CommentSection comments={[]} />);

    expect(
      getByText('There are no comments yet. Be the first to review')
    ).toBeInTheDocument();
  });
});

describe('full feedback section', () => {
  it('should render product comments', async () => {
    const { asFragment } = render(
      <CommentSection comments={feedbackDto.splice(0, 2)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
