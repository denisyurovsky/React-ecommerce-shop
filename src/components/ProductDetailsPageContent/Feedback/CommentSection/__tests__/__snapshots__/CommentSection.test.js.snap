// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`full feedback section should render product comments 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
    data-testid="comments"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-tbqmd1-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root comment css-46bh2p-MuiCardContent-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-1b3y78w-MuiTypography-root"
        >
          Fedor (24.12.2021)
        </h5>
        <span
          aria-label="5 Stars"
          class="MuiRating-root MuiRating-sizeSmall MuiRating-readyOnly css-17cjc3j-MuiRating-root"
          role="img"
        >
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
        </span>
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Cool stuff
        </p>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-tbqmd1-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardContent-root comment css-46bh2p-MuiCardContent-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-1b3y78w-MuiTypography-root"
        >
          Petr (28.12.2021)
        </h5>
        <span
          aria-label="3 Stars"
          class="MuiRating-root MuiRating-sizeSmall MuiRating-readyOnly css-17cjc3j-MuiRating-root"
          role="img"
        >
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
          </span>
          <span>
            <span
              class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
          </span>
        </span>
        <p
          class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
        >
          Awful thing
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
